File: solucion de funciones/bisection/metodoBiseccion.py
import numpy as np
import matplotlib.pyplot as plt



coef  = [1,4,0,-10]  ; p1 = np.poly1d(coef)

x = []
for i in range (1,11):
    x.append(i)

funcion =  p1(x)               # evaluacion del polinomio
# _________________ / / _________________________________

inferior = 1  ; superior = 2   # intervalo de evaluacion
inferior = float(inferior) ; superior = float(superior)
dato = 0                       # dato a buscar
bandera = False

i = 0

def printf(format, *values):
    print(format % values )

while inferior <= superior :
    mitad = float(( inferior + superior ) / 2 )

    if p1(mitad) == dato :
        bandera = True
        break

    if p1(mitad)  > dato :
        superior = mitad
        mitad = float (inferior + superior) / 2

    if p1(mitad)  < dato :
        inferior = mitad
        mitad = float (inferior + superior) / 2   # al implementarlo se realiza la operacion con menos numero de iteraciones

    i += 1

if bandera == True:
    # print( "valor x de solucion: %d  , en la iteracion :  %d" % (mitad,i))
    printf("valor x de solucion: %f  , en la iteracion :  %d", float (mitad) , i)



plt.figure(1)
plt.subplot(111)
plt.plot(x, funcion, label=' funcion ');
#plt.plot(float(mitad) , 'ro', label='punto final')

plt.xlabel(' x ') ; plt.ylabel(' f(x) ')
plt.title(' metodo biseccion ')
plt.grid(True)

plt._show()

# iteracion vs puntos de proceso

# https://www.geeksforgeeks.org/numpy-polyval-in-python/
# https://stackoverflow.com/questions/37352098/plotting-a-polynomial-using-matplotlib-and-coeffiecients

File: solucion de funciones/bisection/metodoBiseccion.m

clc , clear all , close all
%%
% *Metodo Biseccion*
%%
% usando funcion polyval(variable, punto) haciendo uso de busqueda binaria



coef = [1 1 0]; % construir coeficientes polinomio  x^2 + x

coef1 = [1 4 0 -10]; % polinomio  x^3 + 4x^2 -10
x = 1:10;
func = polyval(coef1, x) ;  % entrega una arreglo de los coeficientes evaluados en los vlaores presentes en x

inferior = 1; superior = 2 ;  % intervalo
dato = 0;    % dato corresponde  al valor que se desea encontrar
bandera = 'F';

i = 0 ; 

while(inferior <= superior)
    mitad = (inferior + superior) / 2;
    
    if( polyval(coef1, mitad ) == dato)
        bandera = 'V';                % si la mitad coincide con el dato
        break   
    end
    
    if( polyval(coef1, mitad ) > dato) % se cambia la variable superior 
        superior = mitad;
        mitad = (inferior + superior) / 2;  
    end
    if( polyval(coef1, mitad ) < dato)
        inferior = mitad;
        mitad = (inferior + superior) / 2;  
    end
    
    i = i + 1 ;
end
if (bandera == 'V')
    fprintf(' %s es el valor x de solucion:  ,  en la iteracion : %8.3f  \n ' , mitad , i)
end



% https://www.mathworks.com/help/matlab/ref/fprintf.html
% https://www.mathworks.com/help/matlab/ref/polyval.html
% https://www.youtube.com/watch?v=pxh4QFzDh-Q